Core/clim-core/redefinition.lisp

```diff
@@ -0,0 +1,177 @@
+                         (event  frame-class-redefined-event))
+  ;; We try to preserve the frame size. Store the current size before
+  ;; we make any changes.
+  (multiple-value-bind (width height) (bounding-rectangle-size
+                                       (frame-top-level-sheet client))
+    (setf (frame-command-table client) (new-command-table event)
+          (%frame-layouts client)      (new-layouts event))
+    (let ((new-layout (first (frame-all-layouts client))))
+      ;; Changing the layout involves signaling FRAME-LAYOUT-CHANGED
+      ;; which performs a non-local exit out of this method.
+      (if (eq (frame-current-layout client) new-layout)
+          (when-let ((frame-manager (frame-manager client)))
+            (generate-panes frame-manager client)
+            (layout-frame client width height)
+            (signal 'frame-layout-changed :frame client))
+          (unwind-protect ; Call LAYOUT-FRAME despite non-local exit.
+               (setf (frame-current-layout client) new-layout)
+            (layout-frame client width height))))))
+
+;;; `redefinition-updates-instances-class'
+;;;
+;;; A metaclass for `application-frame' subclasses that update their
+;;; instances when redefined.
+;;;
+;;; Such a subclasses tracks created instances via a list of weak
+;;; pointers stored in the class. When the class is redefined, a
+;;; `frame-class-redefined-event' is dispatched to each frame in that
+;;; list (unless it has become garbage in the meantime).
+
+(defclass redefinition-updates-instances-class (standard-class)
+  ((%instances :accessor instances
+               :initform '())))
+
+(defmethod c2mop:validate-superclass
+    ((class      redefinition-updates-instances-class)
+     (superclass standard-class))
+  t)
+
+(defmethod shared-initialize :before
+    ((instance   redefinition-updates-instances-class)
+     (slot-names t)
+     &key (direct-superclasses nil direct-superclasses-supplied-p))
+  (when direct-superclasses-supplied-p
+    (unless (some (alexandria:rcurry #'subtypep 'application-frame)
+                  direct-superclasses)
+      (error "~@<Instances of ~S must be subclasses of ~S.~@:>"
+             'redefinition-updates-instances-class 'application-frame))))
+
+(defmethod reinitialize-instance :after
+    ((instance redefinition-updates-instances-class) &key)
+  ;; Trigger redefinition notification protocol which will directly or
+  ;; call `handle-frame-class-redefinition' or schedule it for later.
+  (note-frame-class-redefined instance))
+
+(defmethod make-instance :around ((class redefinition-updates-instances-class)
+                                  &key)
+  ;; Store a weak pointer to the created instance for updating it when
+  ;; CLASS is redefines.
+  (let ((instance (call-next-method)))
+    (push (tg:make-weak-pointer instance) (instances class))
+    instance))
+
+(defmethod handle-frame-class-redefinition
+    ((class redefinition-updates-instances-class))
```

Comment:
I wonder if this method won't gain on readability if we get rid of superfluous local functions, i.e like in this sketch?

```
(defun xxx ()
  (if-let ((instances (member-if #'tg:weak-pointer-value (instances class))))
    (let* ((remaining         nil)
           (initargs          (c2mop:class-default-initargs class))
           (new-layouts       (find :layouts initargs :key #'first))
           (new-command-table (find :command-table initargs :key #'first)))
      (loop for weak-instance in instances
            do
               ;; Update instances that haven't become garbage.
               (when-let ((live-instance (tg:weak-pointer-value weak-instance)))
                 (push weak-instance remaining)
                 (with-simple-restart (continue "~@<Skip updating frame ~A~@:>"
                                                instance)
                   (event-queue-append
                    (frame-event-queue live-instance)
                    (make-instance 'frame-class-redefined-event
                                   :sheet             live-instance
                                   :new-layouts       new-layouts
                                   :new-command-table new-command-table))))
            finally
               (setf (instances class remaining))))
    (setf (instances class) nil)))
```

mind that I've dropped typep as per another remark